casper/src/main/scala/coop/rchain/casper/Casper.scala

```diff
@@ -190,74 +183,78 @@ sealed abstract class MultiParentCasperInstances {
+
+          val isValid = maybeCheckPoint.traverse(_ => {
+            val dag = _blockDag.get
+            for {
+              bn <- Validate.blockNumber[F](block, dag)
+              p  <- Validate.parents[F](block, genesis, dag)
+            } yield bn && p
+          })
+
+          isValid.map(_.flatMap {
+            case true =>
+              _blockDag.update(bd => {
+                //add current block as new child to each of its parents
+                val newChildMap = bParents.foldLeft(bd.childMap) {
+                  case (acc, p) =>
+                    val currChildren = acc.getOrElse(p, HashSet.empty[BlockHash])
+                    acc.updated(p, currChildren + hash)
+                }
+
+                bd.copy(
+                  blockLookup = bd.blockLookup.updated(hash, block),
+                  //Assume that a non-equivocating validator must include
+                  //its own latest message in the justification. Therefore,
+                  //for a given validator the blocks are guaranteed to arrive in causal order.
+                  latestMessages = bd.latestMessages.updated(block.sender, hash),
+                  childMap = newChildMap
                 )
-              //TODO: check if block is an equivocation and update observed faults
-
-              maybeCheckPoint.map(_ => {
-                _blockDag.update(bd => {
-                  //add current block as new child to each of its parents
-                  val newChildMap = bParents.foldLeft(bd.childMap) {
-                    case (acc, p) =>
-                      val currChildren = acc.getOrElse(p, HashSet.empty[BlockHash])
-                      acc.updated(p, currChildren + hash)
-                  }
-
-                  bd.copy(
-                    blockLookup = bd.blockLookup.updated(hash, b),
-                    //Assume that a non-equivocating validator must include
-                    //its own latest message in the justification. Therefore,
-                    //for a given validator the blocks are guaranteed to arrive in causal order.
-                    latestMessages = bd.latestMessages.updated(b.sender, hash),
-                    childMap = newChildMap
-                  )
-                })
-
-                true
               })
-            }
+              Some(true)
+
+            case false => None
           })
-          .flatMap {
-            case t @ Some(true) =>
-              //Add successful! Send block to peers and create tuplespace checkpoint
-              CommUtil.sendBlock[F](block) *> Monad[F].pure[Option[Boolean]](t)
-
-            //TODO: Ask peers for missing parents/justifications of blocks
-            case f @ Some(false) => Monad[F].pure[Option[Boolean]](f)
-
-            case _ =>
-              Log[F].info(
-                s"CASPER: Did not add invalid block ${PrettyPrinter.buildString(block.blockHash)}") *>
-                Monad[F].pure[Option[Boolean]](None)
-          }
+        }
+      }.flatMap {
```

Comment:
Can we stick to the for yield syntax? I just find it more elegant and I think its more consistent with @rabbitonweb 's style

```
      //invalid blocks return None
      private def attemptAdd(block: BlockMessage): F[Option[Boolean]] =
        for {
          success <- {
            val hash     = block.blockHash
            val bParents = parents(block)

            if (bParents.exists(p => !_blockDag.get.blockLookup.contains(p))) {
              //Cannot add a block if not all parents are known,
              //but can re-try in the future
              Monad[F].pure[Option[Boolean]](Some(false))
            } else if (block.justifications.exists(j =>
                         !_blockDag.get.blockLookup.contains(j.latestBlockHash))) {
              //Cannot add a block if not all justifications are known,
              //but can re-try in the future
              Monad[F].pure[Option[Boolean]](Some(false))
            } else {
              val Right((maybeCheckPoint, _)) =
                checkpoints.mapAndUpdate[(Option[Checkpoint], Map[ByteString, Checkpoint])](
                  InterpreterUtil.validateBlockCheckpoint(
                    block,
                    genesis,
                    _blockDag.get,
                    storageLocation,
                    storageSize,
                    _
                  ),
                  _._2
                )

              val isValid = maybeCheckPoint.traverse(_ => {
                val dag = _blockDag.get
                for {
                  bn <- Validate.blockNumber[F](block, dag)
                  p  <- Validate.parents[F](block, genesis, dag)
                } yield bn && p
              })

              isValid.map(_.flatMap {
                case true =>
                  _blockDag.update(bd => {
                    //add current block as new child to each of its parents
                    val newChildMap = bParents.foldLeft(bd.childMap) {
                      case (acc, p) =>
                        val currChildren = acc.getOrElse(p, HashSet.empty[BlockHash])
                        acc.updated(p, currChildren + hash)
                    }

                    bd.copy(
                      blockLookup = bd.blockLookup.updated(hash, block),
                      //Assume that a non-equivocating validator must include
                      //its own latest message in the justification. Therefore,
                      //for a given validator the blocks are guaranteed to arrive in causal order.
                      latestMessages = bd.latestMessages.updated(block.sender, hash),
                      childMap = newChildMap
                    )
                  })
                  Some(true)

                case false => None
              })
            }
          }
          result <- success match {
                     case t @ Some(true) =>
                       //Add successful! Send block to peers and create tuplespace checkpoint
                       CommUtil.sendBlock[F](block) *> Monad[F].pure[Option[Boolean]](t)

                     //TODO: Ask peers for missing parents/justifications of blocks
                     case f @ Some(false) => Monad[F].pure[Option[Boolean]](f)

                     case _ =>
                       Log[F].info(
                         s"CASPER: Did not add invalid block ${PrettyPrinter.buildString(block.blockHash)}") *>
                         Monad[F].pure[Option[Boolean]](None)
                   }
        } yield result
```

Reply:
We probably need to break up this function soon too.